crates/bevy_render/Cargo.toml

```diff
@@ -54,3 +54,5 @@ hexasphere = "6.0.0"
 parking_lot = "0.11.0"
 regex = "1.5"
 copyless = "0.1.5"
+## For correct generation of vertex tangents
+mikktspace = { git = "https://github.com/gltf-rs/mikktspace", rev = "a96883cf", default-features = false, features = ["glam"] }
```

Comment:
I would be reluctant to add a dependency to this crate in Bevy. They don't seem to update often, and as they also depend on glam it could make updates harder / slower

Reply:
So you would prefer that we make our own crate that is a reimplementation of it?

Reply:
Not sure...
We could:
* contact the maintainer and discuss with them what's their plan. Bevy seems to be the biggest download driver from their crates (https://crates.io/crates/gltf and https://crates.io/crates/mikktspace). I don't want to impose burden only on their shoulder, would they be ok with help maintaining them?
* vendor it, the license is compatible
* recode it in Bevy... it seems everything is in one file called `generated.rs` with a warning that it's pending a refactor... for the last 3 years

Reply:
My strong preference is to recode it: there's a clear spec to follow, keeping the dependency count (even vendored dep count) is very good, and the code quality is not impressive. 

That is a *very* large amount of unsafe for an algorithm that fundamentally doesn't seem to need any.

Reply:
Some discussion around this happened on Discord in the rendering-dev channel. The doubts raised about using the mikktspace crate are:
- it is not actively maintained
- it contains a lot of unsafe code for something that does not need to use unsafe
- it uses the mikktspace name which could possibly be construed as misrepresentation even if it is a port to another language and it states that
- the zlib license is unclear about what a ‘source distribution’ is and it is unclear whether transpiling or very close language porting would be considered a ‘source distribution’

As such, it seems like we would not want to use the mikktspace crate as-is. We would want an implementation of the Morten S. Mikkelson tangent calculation algorithm that:
- uses safe, idiomatic Rust
- covers the possible interpretations of the zlib license to ensure:
  - the reimplementation does not misrepresent itself as the original, by stating it is not the original and by using a different name
  - covering the possibility that it may be considered as a source distribution if the code is similar enough even if it is in a different programming language, by stating that it is based on the mikktspace implementation (if it were) and referring to the copyright and license of that code
- it may then also make sense to license it as zlib as well

As for what to do for now, there are simple ways of calculating vertex tangents that do not consider all the problem cases addressed by the Mikkelson algorithm and so I’ll probably rework this to calculate tangents in the way specified in the Foundations of Game Engine Development, Volume 2: Rendering book by Eric Lengyel.

Reply:
Yeah looking at the crate, it looks like `generated.rs` is all transpiled and `lib.rs` is a very thin "safe" interface on top. Honestly the unsafety looks _very_ tame, especially given that it is:
1. fully abstracted over in the public interface (which is a safe interface speaking in fixed length float arrays and indices). I tried coming up with a malicious way to call the api and the best I could come up with is implementing num_faces in a way that doesn't match up with the actual length of the underlying data storage. but we'd implement this trait using Vecs, which would panic in that case.
2. generated from code used _widely_ in production (godot, unreal, blender, and unity use mkktspace). if the internals have underlying safety issues, the entire games / graphics ecosystem is in trouble.

Ultimately, I do think it makes sense to rewrite this because unsafety isn't great and this is a relatively small + scoped piece of code. But using mkktspace now seems like a pretty big win from my perspective, given how widely it is used in the industry.

Imo we should just shore up the licensing issue (adding the original license to the generated file with a comment explaining the situation seems reasonable to me), use mikktspace now, and then port it when time allows.

Reply:
So your proposal is to make PRs to the upstream crate to update glam and add the explanation and original license text to the transpiled file?

Reply:
Yup: make a pr to update glam and add the original license text to the transpiled file (with a small comment above that explaining the situation). If they don't accept the PR in a reasonable amount of time, we can just "vendor' the crate.

Reply:
Unless they are _really_ on the ball when it comes to cutting new releases, I think we'll probably want to vendor anyway to make it easy for us to stay up to date on glam deps. Biggest downside there is that it increases the amount of unsafe "under our umbrella". This makes it easier for "scan for unsafe and then yell about it" types to write us off.

Reply:
But yeah whatever. They'd do that even without pulling `mikktspace` under our umbrella.

Reply:
about the original crate pushing releases: @jakobhellermann pushed a pr that was merged at the time of #1795, so 10 months ago, that hasn't been part of a release yet. glam support is still git only, which is why this is a git dependency

Reply:
then lets just vendor in this pr and call it a day.

Reply:
So, for vendoring, I have some thoughts:
- fork the mikktspace crate repo to the bevy organisation to ensure the commit history is not lost but maybe lock it for changes and add one change to the readme noting it is a fork of the original repo for the purpose of preserving history or so
- create `crates/bevy_mikktspace`- I think calling it bevy_mikktspace should be sufficient for making it clear it isn't the original
- copy the files from the forked mikktspace at a specific commit into bevy_mikktspace and note the reference commit in the commit message - we don't seem to like / use git submodules so I guess this is how crevice was vendored
- Update the glam dependency
- Add the original license text into the transpiled file and add an explanation
- Add it to the workspace etc and use it

How does that sound?

Reply:
I like that strategy; it'll let us further refine it later while preserving credit and history.

Reply:
> fork the mikktspace crate repo to the bevy organisation to ensure the commit history is not lost but maybe lock it for changes and add one change to the readme noting it is a fork of the original repo for the purpose of preserving history or so

You can use `git subtree add` to add the full history of mikktspace into the bevy repo. Or you can add a commit to mikktspace to move the code to the right location and then `git merge` it into the bevy repo. Using `git subtree` allows you to extract changes out of the bevy repo again with history preservation. `git subtree` is used for several projects part of the rust-lang/rust repo but still developed independently. Move + `git merge` was used when merging the core engine of miri into rustc as const eval engine while leaving the miri tool as a separate repo.

Reply:
Is there a good way to preserve "line by line" original authorship without dumping the whole mikktspace history into the bevy repo history? 

Alternatively, we could maintain bevy_mikktspace outside of the main bevy repo (although that would be in contrast to bevy_crevice, which does live in the main repo).

Reply:
I messaged the two maintainers from the [`gltf-rs`](https://github.com/gltf-rs) org in case it's possible to find a solution without taking all the extra steps to fork/vendor

Reply:
a first version of [`mikktspace`](https://crates.io/crates/mikktspace) with `glam` support has been published but with an old version of `glam`, I opened [a pr](https://github.com/gltf-rs/mikktspace/pull/25) to update it